### PR TITLE
Allow locked L1 DMA to write to the EFB

### PR DESCRIPTION
**Pull request by magumagu**
I wouldn't have guessed that anyone would do this, but apparently people who write video codecs love doing weird stuff with the locked L1 cache.

I don't have the game, so this is untested.

### CODE REVIEW

File: Source/Core/Core/HW/MemmapFunctions.cpp [hunk 1]
```diff
@@ -460,6 +466,44 @@ void WriteUnchecked_U32(const u32 var, const u32 address)
...
+		for (u32 i = 0; i < 32 * numBlocks; i+=4)
+		{
+			u32 data = bswap(*(u32*)(Memory::m_pL1Cache + ((cacheAddr + i) & 0x3FFFF)));
+			EFB_Write(data, memAddr + i);
+		}
+		return;
+	}
+
+	const u8* src = Memory::m_pL1Cache + (cacheAddr & 0x3FFFF);
+	u8* dst = Memory::GetPointer(memAddr);
```
Comment (degasus) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: Wouldn't this break all other mmio cache lines?

File: Source/Core/Core/HW/MemmapFunctions.cpp [hunk 2]
```diff
@@ -460,6 +466,44 @@ void WriteUnchecked_U32(const u32 var, const u32 address)
...
+	// TODO: It's not completely clear this is the right spot for this code;
+	// what would happen if, for example, the DVD drive tried to write to the EFB?
+	// TODO This is terribly slow.
+	// Avatar: The Last Airbender (GC) uses this for videos.
+	if ((memAddr & 0x0F000000) == 0x08000000)
+	{
+		for (u32 i = 0; i < 32 * numBlocks; i+=4)
+		{
+			u32 data = bswap(*(u32*)(Memory::m_pL1Cache + ((cacheAddr + i) & 0x3FFFF)));
+			EFB_Write(data, memAddr + i);
```
Comment (degasus) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: Oh yeah, this is slow as hell. We'd block for a cpu-gpu thread roundtrip time for each pixel. Also the API overhead in the backend is huge.

File: Source/Core/Core/HW/MemmapFunctions.cpp [hunk 3]
```diff
@@ -90,6 +90,24 @@ static u32 EFB_Read(const u32 addr)
...
+	int y = (addr >> 12) & 0x3ff;
+
+	if (addr & 0x00400000)
+	{
+		g_video_backend->Video_AccessEFB(POKE_Z, x, y, (u32)data);
+		DEBUG_LOG(MEMMAP, "EFB Z Write %08x @ %i, %i", (u32)data, x, y);
+	}
+	else
+	{
+		g_video_backend->Video_AccessEFB(POKE_COLOR, x, y, (u32)data);
```
Comment (degasus) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: It's a bit out of scope of this PR, but what do you think about merging this two functions into:
g_video_backend->Video_AccessEFB(addr - 0x00400000, data);

btw: The casting to (u32) is redundant.

**Comment by JMC47** [comment]
Fixes Avatar: The Last Airbender on GC.

**Comment by delroth** [comment]
Oh god, some people do that? :/

File: Source/Core/Core/HW/MemmapFunctions.cpp (hunk 1 above)
Comment (magumagu) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: I'll add an explicit check.  Not sure what we're supposed to do in that case, though... isn't MMIO sensitive to the size of writes?

File: Source/Core/Core/HW/MemmapFunctions.cpp (hunk 3 above)
Comment (magumagu) [inline Source/Core/Core/HW/MemmapFunctions.cpp]: You mean, as opposed to explicitly computing x and y?  Sounds reasonable.

I'll fix the redundant cast.

**Comment by magumagu** [comment]
@JMC47 Can you do a bit more testing to make sure this is behaving correctly?  Specifically, do the videos appear different in this PR compared to #1865 or the initial version on my clonewars-hack branch?

**Comment by JMC47** [comment]
@magumagu:  It appears just fine in this build.

**Comment by magumagu** [comment]
Updated; I think this addresses all the comments.

**Comment by tueidj** [comment]
FYI this type of operation is _recommended_ in the SDK as one of the fastest ways of achieving direct EFB writes (the other being using the write pipe with the fifo pointed at the EFB).

**Comment by neobrain** [comment]
This is AFAIK also the only way to use the depth testing functionality for EFB pokes properly, fwiw.

**Comment by tueidj** [comment]
That never made much sense to me. How do you write color and Z values in one write transaction when the corresponding EFB addresses are 0x00400000 bytes apart?

**Comment by neobrain** [comment]
@tueidj I don't really understand it really either, unfortunately :/

**Pull request merged**